### DISCUSSION TITLE
Can the "animation.frametime" property of .mcmeta be modified？

### DISCUSSION THREAD

**Discussion by Koalemon** [MoreMcmeta/core#53] [Help Wanted]
I've done some dynamic PNGS, but the minimum unit of refresh rate is 1tick, which is a very low frequency for me.  I tried to modify them, but found it difficult.  I don't know if you've ever thought about doing this.  Maybe you can give me some guidance while I'm still trying to modify them.

**Comment by soir20** [comment]
Yes, you can modify the `frametime` property as specified in the [animation format guide](https://github.com/MoreMcmeta/core/wiki/User-Docs:-Animation-Format#property-definitions). Can you show what you've tried?

Keep in mind that the file extension needs to be `.moremcmeta` and not `.mcmeta`, and the property name is `frametime` and not `frameTime`.

**Reply by Koalemon** [reply]
I'm sorry I didn't notice that your project already had this feature. I used to try to use a mixin modify ".net. Minecraft. Client. Texture. * ", but found frametime endowed with int properties and by a large number of references, this led to I can't change into 0.5 tick.

**Reply by soir20** [reply]
Yes, MoreMcmeta only updates textures every tick. The main reason it's limited this way is that running at a higher frequency doesn't cause a noticeable improvement for most animations. There's also a performance issue with increasing the frequency. I would assume most animations that need to run at a higher frequency would be higher resolution (like a movie), and updating all those pixels at very high frequencies could severely impact performance. MoreMcmeta processes high resolution animations with multiple threads, but that only helps up to a point.

I'd be curious to hear your reasoning for needing a higher frequency. Was your animation converted from a video that ran at a higher frequency? It would be easier to remove extra frames from the animation than to add support for higher frequency animations to MoreMcmeta.

**Reply by Koalemon** [reply]
So moremcmeta can't support 0.5tick update frequency? In fact, my need is to achieve the same render frequency as blockentity, such as making the world block have a visual effect of enchantment. However, the performance cost of blockentity render is greater than the higher frequency of dynamic textures. So I am looking for a frametime that can implement 0.5tick.

**Reply by Koalemon** [reply]
I'm now exporting a bunch of images based on an enchanted item and using them on the block. Achieved the enchanted effect of the block.

**Reply by soir20** [reply]
> So moremcmeta can't support 0.5tick update frequency?

It's not supported currently. It's possible to implement, but I don't think the benefit of adding this feature outweighs the additional complexity and potential performance issues that would come with it.

> my need is to achieve the same render frequency as blockentity

My understanding is that the render frequency of block entities is not constant. The `float` parameter of the `BlockEntityRenderDispatcher#render(BlockEntity, float, PoseStack, MultiBufferSource)` method is the number of ticks that have passed since the block entity was rendered, so it's not a constant amount.

> making the world block have a visual effect of enchantment

There was an enchantment overlay feature request is similar to what you're looking to do: MoreMcmeta/core#15. There's no estimated time for this feature to be implemented currently, though you can subscribe to notifications for it. But it sounds like you've already figured out a solution.

**Reply by Koalemon** [reply]
Yes, the frequency at which blockEntity are rendered is related to the tps frame rate of the game. There are only two solutions to the World block enchantment. The first is dynamic texture, the advantage is low performance loss, the disadvantage is that the fastest frequency is only 1tick. The second is to attach the blockentity property to the block, using the blockentity rendering method, but the performance loss is large.

**Reply by soir20** [reply]
Another solution could be to use a Mixin at the `HEAD` of `BlockEntityRenderDispatcher#render(BlockEntity, float, PoseStack, MultiBufferSource)`. You can wrap the `MultiBufferSource` so that it returns `VertexConsumer`s of type `VertexMultiConsumer.Double`, passing the original render buffer and another buffer using `RenderType.entityGlint()` to the constructor. This is essentially how vanilla MC renders the enchantment glint.

MoreMcmeta's emissive plugin does [something similar](https://github.com/MoreMcmeta/emissive-plugin/blob/6463f20a87948ebb3c1c75eb8690ca0a7c912c48/common/src/main/java/io/github/moremcmeta/emissiveplugin/mixin/BlockEntityRenderDispatcherMixin.java#L43-L50).

(I'm using names from the official Mojang mappings. I'm not sure what the Yarn names are for these classes.)

**Reply by Koalemon** [reply]
That sounds good, but is it bad enough to cause a lot of computer performance loss?It's like a box.

**Reply by soir20** [reply]
There haven't been any noticeable performance issues with this implementation in the emissive plugin. It would render 2x as many quads for the block entities that are rendered as enchanted, but there's no way around that, no matter what solution you would use.

Just to clarify, are you looking to enchant blocks, block entities, or both? My suggestion would work for block entities. However, I think you could use a similar approach for blocks, probably with a Mixin in the `ModelBlockRenderer`  tessellation methods instead of `BlockEntityRenderDispatcher`.

For regular blocks, I don't think you'd need to call `MultiBufferSource#endBatch(RenderType)`. But if you run into issues where the enchantment overlay won't render, it's a good idea to check the batch ordering in `LevelRenderer#renderLevel(PoseStack, float, long, boolean, Camera, GameRenderer, LightTexture, Matrix4f)`.

**Reply by Koalemon** [reply]
I wanted to add enchantment to the ordinary blocks, you know if you put a lot of boxes there will be caused a significant drop in the computer frame rate(fps), so I didn't use the entity blocks,also for the entity blocks I know how to render them. But ordinary squares can only be dynamically textured, such as a 16*(16*1000) texture png.

**Reply by Koalemon** [reply]
Oh, by the way, I forgot to mention that I just need the visual effects of enchantment!

**Reply by soir20** [reply]
I think you'd be able to add the enchantment effect by using a Mixin to modify the `VertexConsumer` in the `ModelBlockRenderer` or `BlockRenderDispatcher` render methods to be a `VertexMultiConsumer.Double`. One of the parameters to the `VertexMultiConsumer.Double` constructor would be the original buffer, and the other would be a buffer for the enchantment glint render type.

**Reply by Koalemon** [reply]
Actually, I've always had a question. Ordinary blocks are not rendered every frame, so it is feasible to add a dynamic block in blockrender?

**Reply by soir20** [reply]
My understanding is that you'd need to use a block entity if you want quads to be dynamically added/removed, like if a block went from being un-enchanted to being enchanted. You could look at how the block breaking overlay is rendered in `LevelRenderer#renderLevel` for potential ideas if this isn't already implemented in Forge/Fabric libraries. I think GeckoLib adds some model animation capabilities, but I'm not familiar enough with it to say whether it would help you do what you want.

I don't think you'd have an issue with the enchantment texture updating. While the baked quads aren't regenerated every frame, textures should be OpenGL-bound and potentially updated every frame. That's how vanilla animated block textures work.

**Comment by soir20** [comment]
There's also some examples in the [demo resource pack](https://www.curseforge.com/minecraft/texture-packs/moremcmeta-demo) (see `assets/minecraft/textures/entity/bat.png.moremcmeta`).

Also, I'm going to convert this to a discussion thread since it's more of a question.

**Comment by soir20** [comment]
@yedaming12 Pinging you here so you get a notification with a link to this discussion thread
